exporter/signalfxexporter/dimensions/dimclient.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dimensions
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// DimensionClient sends updates to dimensions to the SignalFx API
+// This is a port of https://github.com/signalfx/signalfx-agent/blob/master/pkg/core/writer/dimensions/client.go
+// with the only major difference being deduplication of dimension
+// updates are currently not done by this port.
+type DimensionClient struct {
+	sync.RWMutex
+	ctx           context.Context
+	Token         string
+	APIURL        *url.URL
+	client        *http.Client
+	requestSender *ReqSender
+	// How long to wait for property updates to be sent once they are
+	// generated.  Any duplicate updates to the same dimension within this time
+	// frame will result in the latest property set being sent.  This helps
+	// prevent spurious updates that get immediately overwritten by very flappy
+	// property generation.
+	sendDelay time.Duration
+	// Set of dims that have been queued up for sending.  Use map to quickly
+	// look up in case we need to replace due to flappy prop generation.
+	delayedSet map[DimensionKey]*DimensionUpdate
+	// Queue of dimensions to update.  The ordering should never change once
+	// put in the queue so no need for heap/priority queue.
+	delayedQueue chan *queuedDimension
+	// For easier unit testing
+	now func() time.Time
+
+	// TODO: Look into collecting these metrics and other traces via obsreport
+	DimensionsCurrentlyDelayed int64
+	TotalDimensionsDropped     int64
+	// The number of dimension updates that happened to the same dimension
+	// within sendDelay.
+	TotalFlappyUpdates           int64
+	TotalClientError4xxResponses int64
+	TotalRetriedUpdates          int64
+	TotalInvalidDimensions       int64
+	TotalSuccessfulUpdates       int64
+	logUpdates                   bool
+	logger                       *zap.Logger
+}
+
+type queuedDimension struct {
+	*DimensionUpdate
+	TimeToSend time.Time
+}
+
+// NewDimensionClient returns a new client
+func NewDimensionClient(ctx context.Context, token string, apiURL *url.URL,
+	logUpdates bool, logger *zap.Logger, sendDelay int) *DimensionClient {
```

Comment:
Seems too much arguments. Can we use a params struct? We might want to pass more params later as well. What do you think?

Reply:
Added `DimensionClientOptions` as a parameter.